comms/src/outbound_message_service/outbound_message_service.rs

```diff
@@ -0,0 +1,318 @@
+        dest_node_public_key: &PubKey,
+    ) -> Result<Frame, OutboundError>
+    {
+        if flags.contains(IdentityFlags::ENCRYPTED) {
+            let ecdh_shared_secret = PubKey::shared_secret(
+                &self
+                    .node_identity
+                    .secret_key
+                    .clone()
+                    .ok_or(OutboundError::UndefinedSecretKey)?,
+                &dest_node_public_key,
+            )
+            .to_vec();
+            let ecdh_shared_secret_bytes: [u8; 32] = ByteArray::from_bytes(&ecdh_shared_secret)
+                .map_err(|e| OutboundError::SharedSecretSerializationError(e))?;
+            Ok(chacha20::encode(message_envelope_body, &ecdh_shared_secret_bytes))
+        } else {
+            Ok(message_envelope_body.clone())
+        }
+    }
+
+    /// Generate a signature for the MessageEnvelopeHeader from the MessageEnvelopeBody
+    fn sign_envelope_body<R: Rng + CryptoRng>(
+        &self,
+        message_envelope_body: &Frame,
+        rng: &mut R,
+    ) -> Result<Vec<u8>, OutboundError>
+    {
+        let challenge = Challenge::new().chain(message_envelope_body.clone()).result().to_vec();
+        let nonce = SecKey::random(rng);
+        let signature = SchnorrSignature::<PubKey, SecKey>::sign(
+            self.node_identity
+                .secret_key
+                .clone()
+                .ok_or(OutboundError::UndefinedSecretKey)?,
+            nonce,
+            &challenge,
+        )
+        .map_err(|e| OutboundError::SignatureError(e))?;
+        let mut buf: Vec<u8> = Vec::new();
+        signature
+            .serialize(&mut rmp_serde::Serializer::new(&mut buf))
+            .map_err(|_| OutboundError::SignatureSerializationError)?;
+        Ok(buf.to_vec())
+    }
+
+    /// Handler functions use the send function to transmit a message to a peer or set of peers based on the
+    /// BroadcastStrategy
+    pub fn send<R: Rng + CryptoRng>(
+        &self,
+        _broadcast_strategy: BroadcastStrategy, /* TODO: make use of BroadcastStrategy once the PeerManager has been
+                                                 * implemented */
+        flags: IdentityFlags,
+        message_envelope_body: &Frame,
+        rng: &mut R,
+        dest_node_identity: Arc<NodeIdentity<PubKey, SecKey>>, /* TODO: This field is temporary until routing table
+                                                                * has been implemented */
+    ) -> Result<(), OutboundError>
+    {
+        // Todo: use the BroadcastStrategy to select appropriate peer(s) from PeerManager and then construct and send a
+        // personalised message to each selected peer
+        let selected_node_identities = vec![Arc::clone(&dest_node_identity)]; // TODO: Remove with PeerManager call after PeerManager has been implemented
+
+        for dest_node_identity in &selected_node_identities {
+            // Constructing a MessageEnvelope
+            let message_envelope_body =
```

Comment:
If I am reading it right the flags indicate whether the envelope is to be encrypted? I think it will make more sense to check the flag here and make the decision to encrypt here where all the logic for preparing the message is.

Reply:
Change applied